### PR TITLE
Dynamic monthly

### PR DESCRIPTION
**Pull request by Stokalas**
Only file to look at is MonthlyUpdater.cs

### CODE REVIEW

**Review by ciliakas — CHANGES_REQUESTED** [review]
Good idea and tidy implementation, hovewer some glitches need ironing out, highlighted in the comments

**Comment by Stokalas** [comment]
Implemented changes according to the review, which by the way was very useful and pointed main problems, which has helped to solve some problems a lot quicker 🙂

File: SmartSaver/DataManagement/MonthlyUpdater.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    {
+        private readonly DataFilter _dataFilter;
+        private readonly Data _data;
+
+
+        public MonthlyUpdater(DataFilter dataFilter, Data data)
+        {
+            _dataFilter = dataFilter;
+            _data = data;
+        }
```
Comment (ciliakas) [inline SmartSaver/DataManagement/MonthlyUpdater.cs:18]: If we pass a DataFilter to Monthly Updater, wouldn't it make sense to take the Data object from DataFilter? Of course we'd need to create a getter for the DataFilter class

File: SmartSaver/DataManagement/MonthlyUpdater.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        public void UpdateMonthlyEntries(int userId) //eina per visus
+        {
+            var entriesInc = _dataFilter.GetRecurringIncome();
+            var entriesExp = _dataFilter.GetRecurringExpenses();
+
+            foreach (var entry in entriesInc)
+            {
+                var entryDate = entry.Date;
+                var todayDate = DateTime.Today.Date;
```
Comment (ciliakas) [inline SmartSaver/DataManagement/MonthlyUpdater.cs:27]: I think we could move the todayDate variable from the foreach loops since it's always the same

File: SmartSaver/DataManagement/MonthlyUpdater.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            var nextMonth = entry.Date;//assigning current entry data
+            for (var i = 0; i < (months - 1); i++)// one less since last one has to keep isMonthly = true;
+            {
+                nextMonth = TimeManager.MoveToNextMonth(nextMonth);//Adding new entry for each month according to date difference
+                _data.AddIncome(userId, entry.Amount, entry.Title, nextMonth, false, entry.Importance);
+            }
+            /*Adding last entry, which has to keep isMonthly*/
+            nextMonth = TimeManager.MoveToNextMonth(nextMonth);
+            _data.AddIncome(userId, entry.Amount, entry.Title, nextMonth, true, entry.Importance);
+            _data.EditIncomeItem(entry.Id, false);//Moved here since no point to edit in each cycle rotation
```
Comment (ciliakas) [inline SmartSaver/DataManagement/MonthlyUpdater.cs:54]: Very good change, especially considering at the moment every change involves the database and makes everything take up so much more time.

File: SmartSaver/DataManagement/MonthlyUpdater.cs (hunk 2 above)
Comment (ciliakas) [reply SmartSaver/DataManagement/MonthlyUpdater.cs:27]: Could also move this calculation to TimeManager, so we can reuse it if we ever need to :)

File: SmartSaver/DataManagement/Data.cs [hunk 4]
```diff
@@ -416,12 +417,14 @@ public bool EditExpensesItem(int id, DateTime date)
...
+            var tempLocal = Expenses.FirstOrDefault(x => x.Id == id);
+            if (temp != null && tempLocal != null)
 			{
 				temp.IsMonthly = isMonthly;
-				db.SaveChanges();
+                tempLocal.IsMonthly = isMonthly;
+                db.SaveChanges();
 				return true;
 			}
 			else
```
Comment (ciliakas) [inline SmartSaver/DataManagement/Data.cs:430]: Now the editing seems to be working fine, though it would be nice to add this to every Edit method, though I know everything will be reworked soon so there is not much of a point to do this atm

**Review by ciliakas — APPROVED** [review]
Did some tests and the dynamic monthly adding now works! Also nice job with the improvements to efficiency in UpdateMonthlyIncome.
However I really think we need to clean up our Data class, as it is quite a mess now, with some methods affecting only the entries in the database and some affecting both the database and the local data

File: SmartSaver/DataManagement/MonthlyUpdater.cs (hunk 2 above)
Comment (Stokalas) [reply SmartSaver/DataManagement/MonthlyUpdater.cs]: Both good suggestions

File: SmartSaver/DataManagement/MonthlyUpdater.cs (hunk 1 above)
Comment (Stokalas) [reply SmartSaver/DataManagement/MonthlyUpdater.cs:18]: I can see your point, but I would suggest providing required resources inside the handler, or wherever all these objects are created, to see some kind of "hierarchy", maybe I'm understanding it wrong, but I think this would make it easier to understand. On the other hand it would be still understandable if what you suggest would be inplemented. 
TLDR mixed feelings on that aspect

Comment (ciliakas) [reply SmartSaver/DataManagement/MonthlyUpdater.cs:18]: Fair enough

File: SmartSaver/DataManagement/MonthlyUpdater.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            var entriesInc = _dataFilter.GetRecurringIncome();
+            var entriesExp = _dataFilter.GetRecurringExpenses();
+            var todayDate = DateTime.Today.Date;
+
+            foreach (var entry in entriesInc)
+            {
+                var entryDate = entry.Date;
+                var differenceInMonths = TimeManager.DifferenceInMonths(todayDate, entryDate);
+
+                UpdateMonthlyIncome(entry, differenceInMonths, userId);
```
Comment (ciliakas) [inline SmartSaver/DataManagement/MonthlyUpdater.cs:30]: Maybe there should a an if statement here, not to run the update if the entry month is the same as current month or month in advance, because at the moment I found after some testing that it adds a copy (with IsMonthly = false) to the next month

**Review by ciliakas — CHANGES_REQUESTED** [review]
Jumped the ship to soon and now found new errors :(

File: SmartSaver/utilities/TimeManager.cs [hunk 6]
```diff
@@ -2,10 +2,15 @@
...
-    class TimeManager
+    public static class TimeManager
     {
         private static readonly DateTime TwoMonthsAhead = DateTime.Today.AddMonths(2);
 
+        public static int DifferenceInMonths(DateTime first, DateTime second)
+        {
+            return Math.Abs(((first.Year - second.Year) * 12) + first.Month - second.Month);
+        }
+
```
Comment (ciliakas) [inline SmartSaver/utilities/TimeManager.cs:13]: So I'm still having issues with duplicate entries, and I think it's because the difference in months is always positive, as such even if we have a month in advance, Monthly Updater will still add another entry, even though it shouldn't.

**Review by ciliakas — CHANGES_REQUESTED** [review]
The DifferenceInMonths method is very tidy and works well, however in our case we do care if one date is in advance of the other, because in one case we have to add the entries and in other case we shouldn't.

**Comment by Stokalas** [comment]
Since I liked @ciliakas solution to duplicate bug more, I changed and pushed, since he closed his pull request, so just giving credit 🙂

File: SmartSaver/utilities/TimeManager.cs (hunk 6 above)
Comment (Stokalas) [reply SmartSaver/utilities/TimeManager.cs]: Changed this method to not have absolute value

File: SmartSaver/utilities/TimeManager.cs [hunk 7]
```diff
@@ -2,10 +2,15 @@
...
-    class TimeManager
+    public static class TimeManager
     {
         private static readonly DateTime TwoMonthsAhead = DateTime.Today.AddMonths(2);
 
+        public static int DifferenceInMonths(DateTime laterTime, DateTime earlierTime)
+        {
+            return ((laterTime.Year - earlierTime.Year) * 12) + laterTime.Month - earlierTime.Month;
+        }
+
```
Comment (ciliakas) [inline SmartSaver/utilities/TimeManager.cs:13]: After this change everything seems to be working fine, I tested going a few months back, the current month and month into the future and there's no faults 👍

**Review by ciliakas — APPROVED** [review]
Changing the DifferenceInMonths method seems to have fixed the last bug, everything seems to be working perfectly now. Good job 👍
